infra/proposer.go

```diff
@@ -45,22 +50,29 @@ func CreateProposer(addr string, crypto *Crypto) *Proposer {
 		panic(err)
 	}
 
-	return &Proposer{e: endorser}
+	return &Proposer{e: endorser, addr: addr}
 }
 
-func (p *Proposer) Start(signed, processed chan *Elecments, done <-chan struct{}) {
+func (p *Proposer) Start(signed, processed chan *Elecments, done <-chan struct{}, size int) {
 	for {
 		select {
 		case s := <-signed:
+			//fmt.Println("send to " + p.addr)
 			r, err := p.e.ProcessProposal(context.Background(), s.SignedProp)
 			if err != nil || r.Response.Status < 200 || r.Response.Status >= 400 {
-				fmt.Printf("Err processing proposal: %s, status: %d\n", err, r.Response.Status)
+				fmt.Printf("Err processing proposal: %s, status: %d, addr: %s \n", err, r.Response.Status, p.addr)
+				fmt.Println(r)
 				continue
 			}
-
-			s.Response = r
-			processed <- s
-
+			if s.Responses == nil {
+				s.Responses = make(chan *peer.ProposalResponse, size)
+			}
+			s.Responses <- r
+			if len(s.Responses) == size {
+				//fmt.Println("Sending for block")
+				close(s.Responses)
+				processed <- s
+			}
```

Comment:
it's clever to think of using a channel instead of slice. However this doesn't entirely solve the problem - we are risking closing the channel twice here. Consider following sequence of events:
- go routine A enqueues to channel (size = 1)
- go routine B enqueues to channel (size = 2)
- go routine A calls `len` on channel (**len returns 2**)
- go routine B calls `len` on channel (len returns 2)


i'd say, for the sake of simplicity, we could just use a lock here... or an atomic counter

Reply:
Let me try to find a way here.
Currently we are agree for
if array/map unsafe.
if channel, need considering closing twice etc...

Reply:
The reason I don't what to use a lock here, that's lock will block. I am thinking, if some approach as nodejs? a single thread with callback for mutilpute?

Reply:
lock added.